### PR TITLE
SoundWire: cleanup of 'stream' support

### PR DESCRIPTION
**Pull request by plbossart**
This PR revisits the SoundWire 'sdw_stream' support to split allocation and configuration steps. This is necessary if for example the routines are called multiple times from the hw_params stage. This also helps with better error handling.

### CODE REVIEW

File: drivers/soundwire/stream.c [hunk 1]
```diff
@@ -1388,22 +1406,23 @@ int sdw_stream_add_slave(struct sdw_slave *slave,
...
-	m_rt = sdw_alloc_master_rt(slave->bus, stream_config, stream);
+	m_rt = sdw_master_rt_alloc(slave->bus, stream);
 	if (!m_rt) {
 		dev_err(&slave->dev,
 			"alloc master runtime failed for stream:%s\n",
 			stream->name);
 		ret = -ENOMEM;
 		goto error;
 	}
+	ret =  sdw_master_rt_config(m_rt, stream_config);
```
Comment (bardliao) [inline drivers/soundwire/stream.c]: Although sdw_master_rt_config() always return 0, but to be consistent, should we add
```
if (ret < 0)
        goto error;
```
?

File: drivers/soundwire/stream.c [hunk 2]
```diff
@@ -1843,29 +1843,27 @@ int sdw_stream_add_master(struct sdw_bus *bus,
...
 
-skip_alloc_master_rt:
 	ret = sdw_config_stream(bus->dev, stream, stream_config, false);
 	if (ret)
-		goto stream_error;
-
-	ret = sdw_master_port_alloc(m_rt, num_ports);
-	if (ret)
-		goto stream_error;
+		goto unlock;
```
Comment (bardliao) [inline drivers/soundwire/stream.c:1901]: Shouldn't this be `goto alloc_error`? And line 1856 too.

File: drivers/soundwire/stream.c [hunk 3]
```diff
@@ -1950,35 +1953,39 @@ int sdw_stream_add_slave(struct sdw_slave *slave,
...
+	ret = sdw_slave_port_alloc(slave, s_rt, num_ports);
 	if (ret)
-		goto stream_error;
+		goto alloc_error;
 
-	ret = sdw_config_stream(&slave->dev, stream, stream_config, true);
+	ret =  sdw_master_rt_config(m_rt, stream_config);
 	if (ret)
-		goto stream_error;
+		goto unlock;
```
Comment (bardliao) [inline drivers/soundwire/stream.c:2026]: We allocate slave rt successfully already. So I think we should `goto alloc_error` and same as all the `goto unlock` below.

File: drivers/soundwire/stream.c (hunk 1 above)
Comment (plbossart) [reply drivers/soundwire/stream.c]: yes, this is a miss, thanks @bardliao

File: drivers/soundwire/stream.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/stream.c:1901]: I'll have to revisit this, thanks!

File: drivers/soundwire/stream.c (hunk 3 above)
Comment (plbossart) [reply drivers/soundwire/stream.c:2026]: will revisit, thanks!

File: drivers/soundwire/stream.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/stream.c:1901]: actually no @bardliao, it's intentional, see commit message:

````
This is an invasive patch which modifies the error handling flow, with
cleanups only done when an allocation fails. Configuration failures
only return an error code.
````

File: drivers/soundwire/stream.c (hunk 3 above)
Comment (plbossart) [reply drivers/soundwire/stream.c:2026]: same, this was intentional: we only return an errror code on a config error.

File: drivers/soundwire/stream.c (hunk 1 above)
Comment (plbossart) [reply drivers/soundwire/stream.c]: But this is removed in follow-up patches, so the update in the end looks exactly the same. See new branch

File: drivers/soundwire/stream.c (hunk 2 above)
Comment (bardliao) [reply drivers/soundwire/stream.c:1901]: @plbossart Ok, I may still miss something, and didn't get why we only want to free `m_rt` when `sdw_master_port_alloc` failed, yet. Or the question should be why do we need to free `m_rt` when `sdw_master_port_alloc` failed? The `m_rt` may not even allocated by `sdw_stream_add_master`

Comment (plbossart) [reply drivers/soundwire/stream.c:1901]: Good question @bardliao. I didn't really change this code, this was the same before.
But I agree that I could add a test to only free m_rt if it's allocated here.

Edit: that's actually what I do on the 'Slave' side, 


````c
alloc_error:
	/*
	 * we only cleanup what was allocated in this routine
	 */
	if (alloc_master_rt)
		sdw_master_rt_free(m_rt, stream);
	else if (alloc_slave_rt)
		sdw_slave_rt_free(slave, stream);
````

File: drivers/soundwire/stream.c [hunk 4]
```diff
@@ -1270,6 +1259,17 @@ static int sdw_slave_port_alloc(struct sdw_slave *slave,
 	return 0;
 }
 
+static int sdw_slave_port_is_valid_range(struct device *dev, int num)
+{
+	if (!SDW_VALID_PORT_RANGE(num)) {
+		dev_err(dev,
+			"SoundWire: Invalid port number :%d\n", num);
```
Comment (RanderWang) [inline drivers/soundwire/stream.c]: in one line ? dev_err(dev, "SoundWire: Invalid port number :%d\n", num);

File: drivers/soundwire/stream.c [hunk 5]
```diff
@@ -1663,6 +1673,11 @@ int sdw_deprepare_stream(struct sdw_stream_runtime *stream)
 
 	sdw_acquire_bus_lock(stream);
 
+	if (stream->state == SDW_STREAM_DEPREPARED) {
+		ret = 0;
+		goto state_err;
```
Comment (RanderWang) [inline drivers/soundwire/stream.c:1678]: state_err is not a good name.  can we  do as 
````c
sdw_release_bus_lock(stream);
return 0;
````

Comment (plbossart) [reply drivers/soundwire/stream.c:1678]: state_err is what is already in the code, so might as well use it. I don't like duplication of unlock code, it's better a single flow for unlocks, much easier to track.

No change

File: drivers/soundwire/stream.c (hunk 4 above)
Comment (plbossart) [reply drivers/soundwire/stream.c]: yes, that would probably be better. I'll recheck the line splits, thanks for noticing this.

**Comment by plbossart** [comment]
Added minor changes to address @RanderWang suggestions. Diffs since last time

````diff
diff --git a/drivers/soundwire/stream.c b/drivers/soundwire/stream.c
index ca4d32b866d8..6a708fbc0e98 100644
--- a/drivers/soundwire/stream.c
+++ b/drivers/soundwire/stream.c
@@ -943,8 +943,7 @@ static int sdw_slave_port_alloc(struct sdw_slave *slave,
 static int sdw_slave_port_is_valid_range(struct device *dev, int num)
 {
        if (!SDW_VALID_PORT_RANGE(num)) {
-               dev_err(dev,
-                       "SoundWire: Invalid port number :%d\n", num);
+               dev_err(dev, "SoundWire: Invalid port number :%d\n", num);
                return -EINVAL;
        }
 
@@ -956,7 +955,8 @@ static int sdw_slave_port_config(struct sdw_slave *slave,
                                 struct sdw_port_config *port_config)
 {
        struct sdw_port_runtime *p_rt;
-       int i, ret;
+       int ret;
+       int i;
 
        i = 0;
        list_for_each_entry(p_rt, &s_rt->port_list, port_node) {
@@ -1875,9 +1875,7 @@ int sdw_stream_add_master(struct sdw_bus *bus,
 
        m_rt = sdw_master_rt_alloc(bus, stream);
        if (!m_rt) {
-               dev_err(bus->dev,
-                       "Master runtime alloc failed for stream:%s\n",
-                       stream->name);
+               dev_err(bus->dev, "Master runtime alloc failed for stream:%s\n", stream->name);
                ret = -ENOMEM;
                goto unlock;
        }
@@ -1996,9 +1994,7 @@ int sdw_stream_add_slave(struct sdw_slave *slave,
         */
        m_rt = sdw_master_rt_alloc(slave->bus, stream);
        if (!m_rt) {
-               dev_err(&slave->dev,
-                       "alloc master runtime failed for stream:%s\n",
-                       stream->name);
+               dev_err(&slave->dev, "Master runtime alloc failed for stream:%s\n", stream->name);
                ret = -ENOMEM;
                goto unlock;
        }
@@ -2010,9 +2006,7 @@ int sdw_stream_add_slave(struct sdw_slave *slave,
 
        s_rt = sdw_slave_rt_alloc(slave, m_rt);
        if (!s_rt) {
-               dev_err(&slave->dev,
-                       "Slave runtime alloc failed for stream:%s\n",
-                       stream->name);
+               dev_err(&slave->dev, "Slave runtime alloc failed for stream:%s\n", stream->name);
                alloc_slave_rt = false;
                ret = -ENOMEM;
                goto alloc_error;
````

File: drivers/soundwire/stream.c (hunk 4 above)
Comment (plbossart) [reply drivers/soundwire/stream.c]: fixed

**Comment by plbossart** [comment]
https://sof-ci.01.org/linuxpr/PR3313/build6853/devicetest/?model=TGLU_RVP_SDW&testcase=check-suspend-resume-5 shows a fail. let's redo the tests.

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
https://sof-ci.01.org/linuxpr/PR3313/build6857/devicetest/ 7 NTP errors out of 11 devices. Let's retry

**Comment by plbossart** [comment]
SOFCI TEST

**Review by RanderWang — APPROVED** [review]
LGTM

**Review by bardliao — APPROVED** [review]

**Pull request merged**